File: csharp/main/LiteralUInt64.cs
using System;

namespace Deltix.Luminary
{
	public sealed class LiteralUInt64 : LiteralInteger, ILiteral<UInt64>
	{
		public LiteralUInt64(UInt64 value)
			: base(LiteralKind.UInt64)
		{
			Value = value;
		}

		public UInt64 Value { get; }

		public override LiteralInteger CastTo(IntegralType type)
		{
			switch (type)
			{
				case IntegralType.Int64:
					if (Value > Int64.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'Int64'.");
					return new LiteralInt64((Int64) Value);

				case IntegralType.UInt64:
					return this;

				case IntegralType.Int32:
					if (Value > Int32.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'Int32'.");
					return new LiteralInt32((Int32) Value);

				case IntegralType.UInt32:
					if (Value > UInt32.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'UInt32'.");
					return new LiteralUInt32((UInt32) Value);

				case IntegralType.Int16:
					if (Value > (UInt64) Int16.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'Int16'.");
					return new LiteralInt16((Int16) Value);

				case IntegralType.UInt16:
					if (Value > UInt16.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'UInt16'.");
					return new LiteralUInt16((UInt16) Value);

				case IntegralType.Int8:
					if (Value > (UInt64) SByte.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'Int8'.");
					return new LiteralInt8((SByte) Value);

				case IntegralType.UInt8:
					if (Value > Byte.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'UInt8'.");
					return new LiteralUInt8((Byte) Value);

				default:
					throw new ArgumentOutOfRangeException(nameof(type), type, null);
			}
		}

		public override String ToString()
		{
			return $"{Value}u64";
		}
	}
}

File: java/src/main/java/deltix/luminary/LiteralUInt64.java
package deltix.luminary;

import com.google.common.primitives.UnsignedLongs;

public final class LiteralUInt64 extends LiteralInteger {
    private long value;

    public LiteralUInt64(long value) {
        super(LiteralKind.UINT64);
        this.value = value;
    }

    public long getValue() {
        return value;
    }

    @Override
    public LiteralInteger castTo(IntegralType type) {
        switch (type) {
            case INT8:
                if (value < 0 || value > Byte.MAX_VALUE)
                    throw new ClassCastException("Value '" + UnsignedLongs.toString(value) + "' cannot be safely cast to 'Int8'.");
                return new LiteralInt8((byte) value);

            case UINT8:
                if (value < 0 || value > 0xFF)
                    throw new ClassCastException("Value '" + UnsignedLongs.toString(value) + "' cannot be safely cast to 'UInt8'.");
                return new LiteralUInt8((byte) value);

            case INT16:
                if (value < 0 || value > Short.MAX_VALUE)
                    throw new ClassCastException("Value '" + UnsignedLongs.toString(value) + "' cannot be safely cast to 'Int16'.");
                return new LiteralInt16((short) value);

            case UINT16:
                if (value < 0 || value > 0xFFFF)
                    throw new ClassCastException("Value '" + UnsignedLongs.toString(value) + "' cannot be safely cast to 'UInt16'.");
                return this;

            case INT32:
                if (value < 0 || value > Integer.MAX_VALUE)
                    throw new ClassCastException("Value '" + UnsignedLongs.toString(value) + "' cannot be safely cast to 'Int32'.");
                return new LiteralInt32((int) value);

            case UINT32:
                if (value < 0 || value > Integer.MAX_VALUE)
                    throw new ClassCastException("Value '" + UnsignedLongs.toString(value) + "' cannot be safely cast to 'UInt32'.");
                return this;

            case INT64:
                if (value < 0)
                    throw new ClassCastException("Value '" + UnsignedLongs.toString(value) + "' cannot be safely cast to 'SInt64'.");
                return new LiteralInt64(value);

            case UINT64:
                return this;

            default:
                throw new IllegalArgumentException("Undefined integral type.");
        }
    }

    @Override
    public String toString() {
        return UnsignedLongs.toString(value) + "u64";
    }
}
